File: 백준/Silver/30804. 과일 탕후루/과일 탕후루.py
# 과일 탕후루
import sys
input = sys.stdin.readline

n = int(input())
tang = list(map(int,input().split()))
stack = []
max_length = [0] * n
types = set()
result = 0

for i in range(n):
    cur = tang[i]
    types.add(cur)

    if len(types) > 2:
        cnt = 0
        while stack and stack[-1] == tang[i-1]:
            stack.pop()
            cnt += 1
            
        types = {cur, tang[i-1]}
        stack = [cur]
        max_length[i] = cnt + 1
    else:
        max_length[i] = max_length[i-1] + 1
        stack.append(cur)

print(max(max_length))

File: 백준/Silver/30804. 과일 탕후루/과일 탕후루.java
import java.util.*;
import java.io.*;

public class Main {
    public static void main(String[] args) throws IOException {
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        int n = Integer.parseInt(br.readLine());
        StringTokenizer st = new StringTokenizer(br.readLine());
         
        int[] tang = new int[n];
        for (int i = 0; i < n; i++) {
            tang[i] = Integer.parseInt(st.nextToken());
        }

        int start = -1;
        Set<Integer> types = new HashSet<>();
        int result = 0;

        for (int i = 0; i < n; i++) {
            int cur = tang[i];
            types.add(cur);

            if (types.size() > 2) {
                start = i - 2;
                while (tang[start] == tang[i - 1]) {
                    start--;
                }

                types.clear();
                types.add(cur);
                types.add(tang[i - 1]);
                result = Math.max(result, i - start);
            } else {
                result = Math.max(result, i - start);
            }
        }

        System.out.println(result);
    }
}